src/main/java/hudson/plugins/mercurial/MercurialSCMSource.java

```diff
@@ -39,195 +50,289 @@
+            String subdir, HgBrowser browser, boolean clean) {
         super(id);
-        this.installation = installation;
         this.source = source;
         this.credentialsId = credentialsId;
-        this.branchPattern = branchPattern;
-        this.modules = modules;
-        this.subdir = subdir;
-        this.browser = browser;
-        this.clean = clean;
+        this.traits = new ArrayList<>();
+        if (StringUtils.isNotBlank(branchPattern) && !".*".equals(branchPattern) && !".+".equals(branchPattern)) {
+            traits.add(new RegexSCMHeadFilterTrait(branchPattern));
+        }
+        if (clean) {
+            traits.add(new CleanMercurialSCMSourceTrait());
+        }
+        if (installation != null) {
+            traits.add(new MercurialInstallationSCMSourceTrait(installation));
+        }
+        if (browser != null) {
+            traits.add(new MercurialBrowserSCMSourceTrait(browser));
+        }
     }
 
-    public String getInstallation() {
-        return installation;
+    @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE")
+    @SuppressWarnings({"deprecation", "ConstantConditions"}) private Object readResolve() throws ObjectStreamException {
+        if (traits == null) {
+            traits = new ArrayList<>();
+            if (branchPattern != null) {
+                if (StringUtils.isNotBlank(branchPattern) && !".*".equals(branchPattern) && !".+"
+                        .equals(branchPattern)) {
+                    traits.add(new RegexSCMHeadFilterTrait(branchPattern));
+                }
+            }
+            if (clean) {
+                traits.add(new CleanMercurialSCMSourceTrait());
+            }
+            if (installation != null) {
+                traits.add(new MercurialInstallationSCMSourceTrait(installation));
+            }
+            if (browser != null) {
+                traits.add(new MercurialBrowserSCMSourceTrait(browser));
+            }
+        }
+        return this;
     }
 
-    public String getSource() {
+    public @Nonnull String getSource() {
         return source;
     }
 
-    public String getCredentialsId() {
+    public @Nonnull List<SCMSourceTrait> getTraits() {
+        return Collections.unmodifiableList(traits);
+    }
+
+    public @CheckForNull String getCredentialsId() {
         return credentialsId;
     }
 
-    public String getBranchPattern() {
-        return branchPattern;
+    @DataBoundSetter
+    public void setCredentialsId(@CheckForNull String credentialsId) {
+        this.credentialsId = credentialsId;
+    }
+
+    @DataBoundSetter public void setTraits(@CheckForNull List<SCMSourceTrait> traits) {
+        this.traits = new ArrayList<>(Util.fixNull(traits));
     }
 
-    public String getModules() {
-        return modules;
+    @Deprecated @Restricted(DoNotUse.class) @RestrictedSince("2.0") public String getInstallation() {
```

Comment:
Restrictions let up use plugin version stats to guide removing old code. Depreciation does not force dependent plugins to update.

I am making the path to removing the compatibility methods easier to reach